### ISSUE TITLE
Install and use solidity v0.8

### ISSUE THREAD

**Issue by ebouchut** [bet-no-loss/bet-no-loss#16]
# Install Solidity v0.8 

Solidity v0.8 is the latest version to date.

Source: https://docs.soliditylang.org/en/v0.8.0/installing-solidity.html

# Linux

[Read this](https://docs.soliditylang.org/en/v0.8.0/installing-solidity.html#linux-packages)

## MacOS 

```
# if the latest version is 0.8
    brew install solidity 
# else
    brew install solidity@8
```

If you have other versions of solidity installed
```
brew list  --formula | grep solidity | grep -v '^solidity$'
```

then make the symbolic links point to Solidity programs version 0.8 (`solc`, `solidity-upgrade`, `yul-phaser`), like so:
```
brew link --overwrite  solidity
```

## Test Solidity Version

If `solc --version` display `Version: 0.8` then you are good to go.

## Update the solidity version number in source code and config files

* Update the contracts source code (`contracts/*.sol` and `test/*.sol`) to now use  `pragma solidity 0.8;` 
* Update `truffle-config.js` and set `module.exports.compilers.solc.version` to `0.8`

## Compile your contracts

`truffle compile`

## Handle the breaking changes

https://docs.soliditylang.org/en/latest/080-breaking-changes.html

**Comment by ebouchut** [comment]
As `@openzeppelin/test-helpers` does not support Solidity v0.8 yet, 
we fallback to using `v0.7.6` in the meantime.

*Marking this ticket as `improvement` for now.*

**Comment by ebouchut** [comment]
Below a **checklist** of how I **upgraded the project to Solidity version `0.8.3`** from version`0.7.6`.
Some of them are 

- [x] **Update** the **npm packages**
    - [x] Set aside the name of installed  npm packages in the `dependencies` and `devDependencies` sections
    - [x] uninstall all previously installed packages
        ```
        npm uninstall # paste here the name of all previously installed npm packages
        ```
    - [x] Recursively remove  `node_modules`
        ```
        rm -rf node_modules
        ```
    - [x] Install `truffle` locally as a development package in order not to be tied to global version
        ```
        npm install --save-dev truffle
        ```
    - [x] Reinstall all the packages from the `dependencies` section, we previously uninstalled
        ```
        npm install # paste here the name of the npm packages from the **dependencies** section
        ```
    - [x] Install the npm packages from the `devDependecies` section
        ```
        npm install --save-dev # Paste here the name of the npm packages  from the "devDependencies" section
        ```
 - [x] Edit **`truffle-config.js`**: update the version of solidity used by the compiler under `compiler/solc/version`
     ```
    {
      ...
      compilers: {
        solc: {
          version: "0.8.3",
          ...
    ```
- [x] Update the Solidity version in every **contracts** `contrats/*.sol`
    ```
    pragma solidity 0.8.3;
    ```
- [x] Configure **VsCode** to use Solidity version `0.8.3`
    - Right click in the solidity file  and select **`Solidity: Get Solidity Releases`
    - Open a Solidity file in VsCode
        - Right click in the solidity file and:
            -  select  `Solidity: Change Workspace Compiler Version (Remote)`
            - wait a few seconds
            - select `v0.8.3+commit.8d00100c` (or whatever version you want to use)
- [x] Fix the contracts' code and imports
    - Remove the SafeMath import (no longer needed as v0.8+ now natively handles underflow and  overflow):
        - Remove `import "@openzeppelin/contracts/math/SafeMath";`
        - Remove `using SafeMath for ...` below
        - Replace `.add(n)` with `+ n` and respectively `.sub(n)` with `- n`
    - Update the ReentrancyGuard import as this one moved from `contracts/util`to `contracts/security` 
         Replace:  
         `import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";`  
        with  
        `import "@openzeppelin/contracts/security/ReentrancyGuard.sol";`
- [x] Take into account the [(breaking) changes](https://docs.soliditylang.org/en/latest/080-breaking-changes.html)   like for instance `address.send` where address must now be `payable` a coercition would do the trick

**Comment by ebouchut** [comment]
✅ [Here are the changes brought by this branch](https://github.com/bet-no-loss/bet-no-loss/compare/fcef593...f01f1c4)

Closing this issue.

**Comment by ebouchut** [comment]
One last word, here is a blurb on how to bring these changes back to your feature branch.

## Pre-Requisites

You must be on your feature branch `front-dev` 
    `git checkout front-dev`

The goal is to have a clean working directory before starting this procedure.
If `git status` does not say your working directory is clean,
then first make sure to:
- either commit all your changes to your local branch
- or stash the changes 
    `git stash -u`

Now `git status` should say your working directory is clean.

## Procedure

```
# Make sure to be on the feature branch
git checkout front-dev

# Update the local cache of the remote `origin` repository
git fetch

# Update the local master with the changes from origin/master without having to switch branches beforehand
git fetch origin master:master

# 1- Set aside the commits on your feature branch (`dev-front`)
# 2- Add the incoming changes (commits) from master that are not yet on your feature branch
# 3- then replay your commits at the end of the feature branch
git rebase master
```
